### ISSUE TITLE
faust2csound generates incorrect Csound opcode registration

### ISSUE THREAD

**Issue by gogins** [grame-cncm/faust#648]
The `faust2csound` tool generates incorrect C++ code for Csound opcode registration. The generated code is:
```
extern "C" {
    static OENTRY localops[] = {
        {(char*)sym(OPCODE_NAME), sizeof(dataspace), 0, 7, makeDescription(FAUST_OUTPUTS), makeDescription(FAUST_INPUTS, FAUST_ACTIVES),
            (SUBR)init, NULL, (SUBR)process32bits }
    };
    LINKAGE
```
This code does compile and run, but it produces no audio signal. After correcting the generated code to this:
```
extern "C" {
    static OENTRY localops[] = {
        {(char*)sym(OPCODE_NAME), sizeof(dataspace), 0, 3, makeDescription(FAUST_OUTPUTS), makeDescription(FAUST_INPUTS, FAUST_ACTIVES),
            (SUBR)init, (SUBR)process32bits, NULL }
    };
    LINKAGE
}
```
the C++ code not only runs, but also produces the expected audio signal.

The cause of this bug appears to be that the Csound protocol for registering opcodes has changed. The Csound documentation for this is out of date and incorrect, but the current protocol is correctly documented in comments in the source code here: https://github.com/csound/csound/blob/develop/Engine/entry1.c.

My system:
```
mkg@xenakis:~/csound-extended/Opcodes/faust_piano$ uname -a
Linux xenakis 5.4.0-81-generic #91-Ubuntu SMP Thu Jul 15 19:09:17 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
mkg@xenakis:~/csound-extended/Opcodes/faust_piano$ csound --version
time resolution is 1000.000 ns
0dBFS level = 32768.0
--Csound version 6.17 beta (double samples) Sep  3 2021
[commit: 29f907049413cbf5a0adfe927cc7437cac322fac]
libsndfile-1.0.28
```

More details, and test code, may be found at https://github.com/gogins/csound-extended/tree/master/Opcodes/faust_piano.

**Comment by sletz** [comment]
Are you use you are using a recent version? The current code here: https://github.com/grame-cncm/faust/blob/master-dev/architecture/csound.cpp#L239 uses the correct code AFAICS.

**Comment by gogins** [comment]
I cloned the Faust repository and rebuilt from the "master-dev" branch with
"make developer". There were build errors, which I corrected. After "sudo
make install" I re-ran my test code and everything worked fine using faust
version 2.34.4 without having to patch the generated code.

To get the build to complete I had to do this:

***@***.***:~/faust$ git diff
diff --git a/compiler/generator/llvm/llvm_dsp_aux.hh
b/compiler/generator/llvm/llvm_dsp_aux.hh
index 8a980d563..226eb0686 100644
--- a/compiler/generator/llvm/llvm_dsp_aux.hh
+++ b/compiler/generator/llvm/llvm_dsp_aux.hh
@@ -52,7 +52,7 @@
 #define MovePTR(ptr) std::move(ptr)
 #define PASS_MANAGER legacy::PassManager
 #define FUNCTION_PASS_MANAGER legacy::FunctionPassManager
-#define sysfs_binary_flag sys::fs::F_None
+#define sysfs_binary_flag sys::fs::OF_None
 #define OwningPtr std::unique_ptr
 #define llvmcreatePrintModulePass(out) createPrintModulePass(out)
 #define GET_CPU_NAME llvm::sys::getHostCPUName().str()
diff --git a/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp
b/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp
index 2c4e2211e..740a4440e 100644
--- a/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp
+++ b/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp
@@ -423,7 +423,7 @@ bool
llvm_dynamic_dsp_factory_aux::writeDSPFactoryToObjectcodeFileAux(const stri
     fModule->setDataLayout(TheTargetMachine->createDataLayout());

     error_code EC;
-    raw_fd_ostream  dest(object_code_path.c_str(), EC, sys::fs::F_None);
+    raw_fd_ostream  dest(object_code_path.c_str(), EC, sys::fs::OF_None);
     if (EC) {
         errs() << "ERROR : writeDSPFactoryToObjectcodeFile could not open
file : " << EC.message();
         return false;


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by sletz** [comment]
Thanks. Which OS and LLVM version are you using ?

**Comment by sletz** [comment]
Fixed in https://github.com/grame-cncm/faust/commit/513a5c7f5b0126e18c9dac95f7b85cd2bd3d4c65

**Comment by gogins** [comment]
llvm -13 and Ubuntu 20.04.

Regards,
Mi8ke

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by gogins** [comment]
I had trouble compiling Faust from sources on my system, so installed the
system package for Faust.

Regards,
Mike

**Comment by gogins** [comment]
One last email about my issues with Faust and Csound to clarify things.

I couldn't build a current version of Faust from source code on Ubuntu
20.04, so I had to use the Faust 2.2.0 system package for Ubuntu Linux.

With this, I was able to build and use the "faustgen" family of Csound
opcodes, which use LLVM to compile Faust code at run time.

As I mentioned, the code that "faust2csound" from Faust 2.2.0 generates for
registering a plugin Csound opcode is not correct for the current version
of Csound.

Fortunately, this proved very easy to fix by patching one line in
"csound.cpp" architecture file in Faust 2.2.0 to use Csound's current
protocol for opcode registration.

With this minor change, both Csound's "faustgen" opcodes and Faust's
"faust2csound" tool seem to work perfectly.

Details are here:
https://github.com/gogins/csound-extended/tree/master/Opcodes/faust_piano.

Regards,
Mike


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com
